### PR TITLE
nixos-rebuild: Print commands as they are run on verbose flag

### PR DESCRIPTION
**Pull request by Profpatsch**
nixos-rebuild would forward all verbose flags to nix, but not actually
increase its own verbosity.
Which would make it hard to see what’s going on without resorting to
tools like strace or adding `set -x` to the script.

We add a simple exec wrapper around the most “interesting” commands,
that is nix tool invocations. If any verbosity flag is given, it will
set the verbosity flag of the nixos-rebuild script all commands
wrapped into `runCmd` will start logging their invocation.

This is done via a `logVerbose` function, which can also be used to
print logging messages when in verbose mode, to aid debugging why
nixos-rebuild does what it does. A few messages are added, but
potentially we could log a lot more details.

Now, testing all of the changed code paths turns out to be rather
hard, so I am unsure how to verify that for all changed commands
behaviour stays indeed the same.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Profpatsch** [comment]
I can only test it on my local machine, trying to hit a few codepaths, but I don’t know if there is any testsuite that tests more of `nixos-rebuild`?

**Comment by Profpatsch** [comment]
Based on some ideas by @aszlig I added all installer tests to the test closure, which should give us a good integration test that at least nothing broke that had an existing test.

**Comment by Profpatsch** [comment]
@GrahamcOfBorg test nixos-rebuild

**Comment by Profpatsch** [comment]
Hm, maybe all of these tests are a bit too much for ofborg …

**Comment by Profpatsch** [comment]
I put only a single installer test, and now ofborg seems to be happy. So I’m reasonably sure nixos-rebuild still works for the most common case.

**Comment by Profpatsch** [comment]
Added Github suggestions to reviewer list.

File: pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh [hunk 1]
```diff
@@ -516,7 +542,7 @@ fi
 # If we're not just building, then make the new configuration the boot
 # default and/or activate it now.
 if [[ "$action" = switch || "$action" = boot || "$action" = test || "$action" = dry-activate ]]; then
-    if ! targetHostCmd "$pathToConfig/bin/switch-to-configuration" "$action"; then
+    if ! runCmd targetHostCmd "$pathToConfig/bin/switch-to-configuration" "$action"; then
```
Comment (thiagokokada) [inline pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:545]: Will this actually work? I am not seeing you return the error code here: https://github.com/NixOS/nixpkgs/pull/163346/files#diff-fc23c722779993fcffc932965cd143b265aa196d577ca10173b63e5bdcbc0001R161

Comment (Profpatsch) [reply pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:545]: It should work just fine:

```
# foo() { "$@"; }
# foo false
# $?
1: command not found
```

**Review by trofi — APPROVED** [review]
Built local system successfully with a few runs of `# nixos-rebuild switch --verbose -I nixpkgs=$PWD`.

**Comment by Profpatsch** [comment]
Sorry for the messy force push, I removed `runCmd` from two places where it would only print the name of an internal function instead of a real command.

**Comment by Profpatsch** [comment]
I’d merge soon if there are no counterpoints.

**Pull request merged**